Add trigger 'htmx:sseClose' to sse close event

sondalex opened the pull request:
This PR adds a new trigger, `"htmx:sseClose"`, which is emitted when an SSE closing message is received.

Telroshan commented:
Hey, while I like the idea, it seems you aren't firing this event in all cases where a source might be closed, was it intended?
For ex I can see those places in the code where you didn't apply it:
https://github.com/bigskysoftware/htmx-extensions/blob/8845604994bdae5b5dec0d5510732fd89f0a700c/src/sse/sse.js#L47-L49

https://github.com/bigskysoftware/htmx-extensions/blob/8845604994bdae5b5dec0d5510732fd89f0a700c/src/sse/sse.js#L239-L243

Also, do you think that, on error, a `sseClose` event should also be fired? (since it can close the event source, cf the reconnection algorithm)

Also, to accept this PR it'd be good to add the following :
- Documentation for the event, as a recent PR (#25) added documentation for all the events dispatched by this extension, we should keep it up to date
- Add test cases to ensure the event is correctly fired in all closing situations

sondalex commented:
Hi, thank you for the review and recommendations! I understand your concerns about not firing the `htmx:sseClose` event in all cases where a source might be closed.

Initially, I intended to trigger the event only when a closing message is received, not when the connection is closed due to the removal of the element containing `sse-connect`. However, I see the value in firing the event in all closing situations.

Regarding your suggestion to fire the same event for all closing connections, I'm open to implementing `htmx:sseClose` for all cases. Alternatively, I could introduce separate events, such as `htmx:sseNodeClose` for node replacement and `htmx:sseMessageClose` for closing messages. What are your thoughts on this ?

Telroshan commented:
Oh I see, makes sense! I think a parameter could make more sense? An enum-like value in the event's details that lets the user know, if it matters to them, what was the reason of this sse close. So that if someone just wants to bind some logic upon a sse connection closing, they don't have to bind 3 different events, and if they need to differentiate the scenarios, they may simply use if/else or a switch statement

sondalex commented:
Hi @Telroshan ,  I have added test cases and implemented the close types in the event details. Thank you for the recommendation.

Users can access the sseClose details as follow:

```javascript
document.body.addEventListener('htmx:sseClose', function (e) {
    const reason = e.detail.type
    switch(reason) {
        case "nodeMissing":
            // Parent node is missing and therefore connection was closed
            ...
        case "nodeReplaced":
            // Parent node replacement caused closing of connection 
            ...
        case "message":
            // connection was closed due to reception of message sse-close 
            ...
    }
})
```

Telroshan commented:
Looks good @sondalex, thanks!
Last request before we can merge this ; could you please remove the formatting changes? To concentrate the diff on the actual code changes

sondalex commented:
Sure @Telroshan , I have just fixed it.

Telroshan commented on src/sse/sse.js:
```diff
@@ -45,6 +45,9 @@ This extension adds support for Server Sent Events to htmx.  See /www/extensions
           // Try to remove remove an EventSource when elements are removed
           if (internalData.sseEventSource) {
+            api.triggerEvent(parent, 'htmx:sseClose', {
+              type: 'nodeReplaced',
```
Is there any reason you would not want to pass `source: internalData.sseEventSource` down the event, as you passed the event source reference in the 2 other event types below?

sondalex replied:
I missed this one, thank you for noticing.

Telroshan commented:
Perfect, thanks!
